pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageRedeliveryController.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service.persistent;
+
+import com.google.common.collect.ComparisonChain;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.bookkeeper.util.collections.ConcurrentLongLongPairHashMap;
+import org.apache.bookkeeper.util.collections.ConcurrentLongLongPairHashMap.LongPair;
+import org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet;
+import org.apache.pulsar.common.util.collections.LongPairSet;
+
+public class MessageRedeliveryController {
+    private final LongPairSet messagesToRedeliver;
+    private final ConcurrentLongLongPairHashMap hashesToBeBlocked;
+
+    public MessageRedeliveryController(boolean allowOutOfOrderDelivery) {
+        this.messagesToRedeliver = new ConcurrentSortedLongPairSet(128, 2);
+        this.hashesToBeBlocked = allowOutOfOrderDelivery ? null : new ConcurrentLongLongPairHashMap(128, 2);
+    }
+
+    public boolean add(long ledgerId, long entryId) {
+        return messagesToRedeliver.add(ledgerId, entryId);
+    }
+
+    public boolean add(long ledgerId, long entryId, long stickyKeyHash) {
+        if (hashesToBeBlocked != null) {
+            hashesToBeBlocked.put(ledgerId, entryId, stickyKeyHash, 0);
+        }
+        return messagesToRedeliver.add(ledgerId, entryId);
+    }
+
+    public boolean remove(long ledgerId, long entryId) {
+        if (hashesToBeBlocked != null) {
+            hashesToBeBlocked.remove(ledgerId, entryId);
+        }
+        return messagesToRedeliver.remove(ledgerId, entryId);
+    }
+
+    public int removeAllUpTo(long markDeleteLedgerId, long markDeleteEntryId) {
+        if (hashesToBeBlocked != null) {
+            for (LongPair longPair : hashesToBeBlocked.keys()) {
```

Comment:
Using `key()` will make a copy of the whole set, so we should instead do the `forEach()`.

Reply:
@merlimat If we try to remove any entry in `forEach()`, it seems to cause a deadlock. Is there any way to avoid this without using `keys()`?

Reply:
@massakam Does the deadlock happen in 2.8.x / master branch? There are changes in #9787 that resolved some dead lock issues.

Reply:
@lhotari It happens in the master branch. `hashesToBeBlocked` is an instance of `org.apache.bookkeeper.util.collections.ConcurrentLongLongPairHashMap`, so it should be unrelated to those fixes.

Reply:
Ok, we can go on with this but we should also import here a fixed version of `ConcurrentLongLongPairHashMap` that doesn't have the problem

Reply:
> but we should also import here a fixed version of `ConcurrentLongLongPairHashMap` that doesn't have the problem

@merlimat Such a version doesn't exist yet, does it?

Reply:
Add the keys to be removed in the `forEach()` loop to the newly created list and remove them from `hashesToBeBlocked` after exiting the loop. This makes the entries to be copied part of the set rather than the whole set. PTAL.